src/impy/models/dpmjetIII.py

```diff
@@ -95,37 +94,27 @@ def __init__(self, evt_kin, *, seed=None):
             assert False, "Unknown DPMJET version, IO common block not detected"
         self._lib.pydat1.mstu[10] = lun
 
-        self.kinematics = evt_kin
-
-        # Relax momentum and energy conservation checks at very high energies
-        if evt_kin.ecm > 5e4:
-            # Relative allowed deviation
-            self._lib.pomdls.parmdl[74] = 0.05
-            # Absolute allowed deviation
-            self._lib.pomdls.parmdl[75] = 0.05
-
         # Prevent DPMJET from overwriting decay settings
         # self._lib.dtfrpa.ovwtdc = False
         # Set PYTHIA decay flags to follow all changes to MDCY
         self._lib.pydat1.mstj[21 - 1] = 1
         self._lib.pydat1.mstj[22 - 1] = 2
-        self._set_final_state_particles()
 
-    def _cross_section(self, kin=None, precision=None):
-        kin = self.kinematics if kin is None else kin
-        # we override to set precision
-        if (kin.p1.A and kin.p1.A > 1) or kin.p2.A > 1:
-            assert kin.p2.A >= 1, "DPMJET requires nucleons or nuclei on side 2."
+    def _cross_section(self, kin, precision=None):
+        self._set_kinematics(kin)
+        assert kin.p2.A >= 1  # should be guaranteed by MCRun._validate_kinematics
+
+        projectile_id = 2212 if kin.p1.is_nucleus else kin.p1
+
+        if (kin.p2.is_nucleus and kin.p2.A > 1) or (kin.p1.is_nucleus and kin.p1.A > 1):
             if precision is not None:
                 saved = self._lib.dtglgp.jstatb
                 # Set number of trials for Glauber model integration
                 self._lib.dtglgp.jstatb = precision
             self._lib.dt_xsglau(
                 kin.p1.A or 1,
                 kin.p2.A or 1,
-                self._lib.idt_icihad(2212)
-                if (kin.p1.A and kin.p1.A > 1)
-                else self._lib.idt_icihad(kin.p1),
+                projectile_id,
```

Comment:
DT_XSGLAU needs bamjet codes. It is also the right routine to call for hadron-nucleus.

Reply:
Oops, I forgot to call idt_icihad on the projectile.

Reply:
Fixed.